### PR TITLE
Adding command to get recent failure log from failed deployment

### PR DESCRIPTION
**Pull request by emmasax1**
Description and Impact
----------------------
When a user runs `get-failure-log [account]`, this will open the most recent failure log from the current stack in a browser.

Deploy Plan
-----------
* checkout master
* `op accept-pull 97`
* `soyuz deploy` and pick RubyGems with a minor release

Rollback Plan
-------------
* To roll back this change, revert the merge with: `git revert -m 1 MERGE_SHA` and perform another deploy.

URLs
----
http://docs.aws.amazon.com/sdkforruby/api/Aws/OpsWorks/Client.html

QA Plan
-------
- [x] Run `bin/opsicle failure-log [account]` with the `.opsicle` file for that stack locally, and then watch the browser open
- [x] Actually navigate in browser to the most recent failure log, and compare to make sure the two logs are the same

### CODE REVIEW

**Comment by emmasax1** [comment]
QA 👍 

screen shot 2016-07-15 at 2 03 20 pm
screen shot 2016-07-15 at 2 03 13 pm

File: .gitignore [hunk 1]
```diff
@@ -15,6 +15,9 @@ spec/reports
 test/tmp
 test/version_tmp
 tmp
+lib/opsicle/.DS_Store
+lib/opsicle/monitor/.DS_Store
```
Comment (chrisarcand) [inline .gitignore]: These lines shouldn't be necessary with the line directly below :)

**Comment by chrisarcand** [comment]
The `get-` prefix seems unnecessary to me. IMO just `failure-log <env>`, rename the class `FailureLog` and add a `#fetch` method containing the logic for fetching the log, called from `#execute`. Shorter, less redundant, and if there's any feature added in the future related to failure logs you now haven't boxed yourself into a corner.

File: lib/opsicle/commands/failure_log.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+      fetch
+    end
+
+    def fetch
+      stack_id_hash = {stack_id: @stack.stack_id}
+      deployments = @client.opsworks.describe_deployments(stack_id_hash).deployments
+      failed_deployments = deployments.select{ |deploy| !deploy.status.eql? "successful" }
+      failed_deployment_id = failed_deployments.first.deployment_id
+      command_list = @client.opsworks.describe_commands(deployment_id: failed_deployment_id)
```
Comment (bbergstrom) [inline lib/opsicle/commands/failure_log.rb]: Not sure I like it picking a server at random for the failed deployment.  Did you consider prompting the user for a list of servers in the failed deployment?  That fits the pattern of other opsicle commands choosing from a list of servers.  Also, with some deployment commands the failure may be different on a particular layer or server so letting the user make a intentional choice here would be a good capability.

File: lib/opsicle/commands/failure_log.rb [hunk 3]
```diff
@@ -15,14 +15,45 @@ def execute
...
+
+      target_failed_command = fetch_target_command(involved_instance_id, failed_deployment_id)
+      log_url = target_failed_command.first.log_url
+      
+      system("open", log_url) if log_url
+    end
+
+    def fetch_failed_deployments
+      deployments = @client.opsworks.describe_deployments(stack_id: @stack.stack_id).deployments
+      return deployments.select{ |deploy| !deploy.status.eql? "successful" }
```
Comment (bbergstrom) [inline lib/opsicle/commands/failure_log.rb]: `return` is extraneous here as the last return value of a method gets returned by default in Ruby.

Also, a more Rubyist way of writing this, which may be more readable:
```
deployments.reject { |deploy| deploy.status.eql? "successful" }
```
http://ruby-doc.org/core-2.2.5/Enumerable.html#method-i-reject

File: lib/opsicle/commands/failure_log.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+      target_failed_command = fetch_target_command(involved_instance_id, failed_deployment_id)
+      log_url = target_failed_command.first.log_url
+      
+      system("open", log_url) if log_url
+    end
+
+    def fetch_failed_deployments
+      deployments = @client.opsworks.describe_deployments(stack_id: @stack.stack_id).deployments
+      deployments.reject{ |deploy| deploy.status.eql? "successful" }
```
Comment (bbergstrom) [inline lib/opsicle/commands/failure_log.rb]: Will this get "running" deployments as well?

File: lib/opsicle/commands/failure_log.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      failed_deployments = fetch_failed_deployments
+      failed_deployment_id = failed_deployments.first.deployment_id
+      failed_deployments_instances = failed_deployments.first.instance_ids
+
+      involved_instance_id = fetch_instance_id(failed_deployments_instances)
+
+      target_failed_command = fetch_target_command(involved_instance_id, failed_deployment_id)
+      log_url = target_failed_command.first.log_url
+      
+      system("open", log_url) if log_url
```
Comment (bbergstrom) [inline lib/opsicle/commands/failure_log.rb]: Consider adding a message to the user if there are no failures in available history.
```
brianbergstrom@bb-mbp-4 opsicle (add-get-failure-log-command)$ bundle exec opsicle failure-log staging
Getting most recent failure log...
brianbergstrom@bb-mbp-4 opsicle (add-get-failure-log-command)$
```

File: lib/opsicle/commands/failure_log.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    def execute
+      puts "Getting most recent failure log..."
+
+      fetch
+    end
+
+    def fetch
+      failed_deployments = fetch_failed_deployments
+      failed_deployment_id = failed_deployments.first.deployment_id
```
Comment (bbergstrom) [inline lib/opsicle/commands/failure_log.rb]: I think you need to add a check here for the existence of a failed deployment.
```
brianbergstrom@bb-mbp-4 opsicle (add-get-failure-log-command)$ bundle exec opsicle --debug failure-log production
Getting most recent failure log...
error: undefined method `deployment_id' for nil:NilClass
/Users/brianbergstrom/projects/opsicle/lib/opsicle/commands/failure_log.rb:19:in `fetch': undefined method `deployment_id' for nil:NilClass (NoMethodError)
	from /Users/brianbergstrom/projects/opsicle/lib/opsicle/commands/failure_log.rb:14:in `execute'
	from /Users/brianbergstrom/projects/opsicle/bin/opsicle:59:in `block (2 levels) in <top (required)>'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/gems/gli-2.14.0/lib/gli/command_support.rb:126:in `call'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/gems/gli-2.14.0/lib/gli/command_support.rb:126:in `execute'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/gems/gli-2.14.0/lib/gli/app_support.rb:296:in `block in call_command'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/gems/gli-2.14.0/lib/gli/app_support.rb:309:in `call'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/gems/gli-2.14.0/lib/gli/app_support.rb:309:in `call_command'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/gems/gli-2.14.0/lib/gli/app_support.rb:83:in `run'
	from /Users/brianbergstrom/projects/opsicle/bin/opsicle:238:in `<top (required)>'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/bin/opsicle:23:in `load'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/bin/opsicle:23:in `<main>'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/bin/ruby_executable_hooks:15:in `eval'
	from /Users/brianbergstrom/.rvm/gems/ruby-2.1.4@opsicle/bin/ruby_executable_hooks:15:in `<main>'
```

**Comment by bbergstrom** [comment]
CRQA 👍

**Pull request merged**